elastic/translate.go

```diff
@@ -1,185 +1,186 @@
 package elastic
 
 import (
-	"encoding/json"
+	"errors"
 	"fmt"
-	"log"
 	"strings"
 )
 
 // Leaf is a terminal node of a query
 // All terminal query terms implement the Translate() method
-type Leaf interface {
-	Translate() map[string]interface{}
+type leaf interface {
+	Translate() Term
 }
 
-func encodeParam(concept string, param map[string]interface{}) map[string]interface{} {
-	operator := param["operator"].(string)
-	var conceptID string
-	if operator == "set" {
-		conceptID = "_id"
+// encodeParam encodes a parameter from a query to its corresponding leaf type
+func encodeParam(concept string, param map[string]interface{}) (map[string]interface{}, error) {
+	o := param["operator"].(string)
+	var cid string
+	if o == "set" {
+		cid = "_id"
 	} else {
-		conceptID = param["id"].(string)
+		cid = param["id"].(string)
 	}
 	var conceptPath string
-	if concept != "" && operator != "set" {
-		conceptPath = concept + "." + conceptID
+	if concept != "" && o != "set" {
+		conceptPath = concept + "." + cid
 	} else {
-		conceptPath = conceptID
+		conceptPath = cid
 	}
-	value := param["value"]
-	var paramLeaf Leaf
-	switch operator {
+	v := param["value"]
+	var l leaf
+	switch o {
 	case "set":
-		paramLeaf = &MemberTerm{conceptPath, value}
+		l = &MemberTerm{conceptPath, v}
 	case "eq":
-		paramLeaf = &EqualityTerm{conceptPath, value, "+"}
+		l = &EqualityTerm{conceptPath, v, false}
 	case "-eq":
-		paramLeaf = &EqualityTerm{conceptPath, value, "-"}
+		l = &EqualityTerm{conceptPath, v, true}
 	case "undefined":
-		paramLeaf = &DefinitionTerm{conceptPath, false}
+		l = &DefinitionTerm{conceptPath, false}
 	case "defined":
-		paramLeaf = &DefinitionTerm{conceptPath, true}
+		l = &DefinitionTerm{conceptPath, true}
 	case "one":
-		sliceValue := value.([]string)
-		paramLeaf = &OneTerm{conceptPath, sliceValue}
+		s := v.([]string)
+		l = &OneTerm{conceptPath, s}
 	case "match":
-		stringValue := value.(string)
-		paramLeaf = &MatchTerm{conceptPath, stringValue}
+		s := v.(string)
+		l = &MatchTerm{conceptPath, s}
 	case "query":
-		stringValue := value.(string)
-		paramLeaf = &QueryTerm{conceptPath, stringValue}
+		s := v.(string)
+		l = &QueryTerm{conceptPath, s}
 	case "gt":
-		paramLeaf = &GreaterThanTerm{conceptPath, value, false}
+		l = &GreaterThanTerm{conceptPath, v, false}
 	case "gte":
-		paramLeaf = &GreaterThanTerm{conceptPath, value, true}
+		l = &GreaterThanTerm{conceptPath, v, true}
 	case "lt":
-		paramLeaf = &LessThanTerm{conceptPath, value, false}
+		l = &LessThanTerm{conceptPath, v, false}
 	case "lte":
-		paramLeaf = &LessThanTerm{conceptPath, value, true}
+		l = &LessThanTerm{conceptPath, v, true}
 	case "range":
-		sliceValue := value.([]interface{})
-		gt := sliceValue[0]
-		lt := sliceValue[1]
-		paramLeaf = &RangeTerm{conceptPath, lt, gt}
+		s := v.([]interface{})
+		gt := s[0]
+		lt := s[1]
+		l = &RangeTerm{conceptPath, lt, gt}
```

Comment:
No need to assign `gt` and `lt`, just pass them in to `RangeTerm{..., s[0], s[1]}`